emily_cron/app/services/deposit_processor.py

```diff
@@ -160,6 +227,65 @@ def update_deposits(self) -> None:
 
         logger.info("Deposit status update job completed")
 
+    def _group_rbf_transactions(self, rbf_txs: list[EnrichedDepositInfo]) -> dict[str, set[str]]:
+        """Groups transactions that belong to the same RBF chain.
+
+        An RBF chain consists of an original transaction and all subsequent
+        transactions that attempt to replace it (or replace a replacement).
+        This function takes a list of transactions known to have replacements
+        and identifies these complete chains.
+
+        The core logic involves iterating through the transactions and merging
+        any sets of transaction IDs (txids) that share common members. For example,
+        if Tx A is replaced by Tx B, and Tx B is replaced by Tx C, they form a
+        single chain {A, B, C}. If we later find Tx D replaces Tx B, Tx D is also
+        added to the same chain {A, B, C, D}.
+
+        Args:
+            rbf_txs: A list of enriched deposit information objects, specifically
+                     those where the `rbf_txids` attribute is non-empty.
+
+        Returns:
+            A dictionary where each key is an arbitrary transaction ID from a
+            discovered RBF chain, and the value is a set containing all
+            transaction IDs belonging to that complete RBF chain.
+        """
+        rbf_groups: dict[str, set[str]] = {}
+
+        # First, build groups of related transactions (original + replacements)
+        for tx in rbf_txs:
+            # Create a set of all txids in this RBF chain
+            chain_txids = set(tx.rbf_txids)
+            chain_txids.add(tx.bitcoin_txid)
+
+            # Find all groups that overlap with this chain
+            overlapping_groups = []
+            for group_id, group_txids in rbf_groups.items():
+                if chain_txids.intersection(group_txids):
```

Comment:
trivia: It looks like Mempool returns the _full_ RBF chain when queried for any transaction along the chain. So here the intersection should in practice always be either full or empty.

Reply:
nit: Based on this information you could actually reduce the complexity of this from N**2 to NlogN by:

1. Storing the RBF chain sorted.
2. Sorting `rbf_txs` based on the chains.

(See my other comment about `itertools.groupby`).